### PR TITLE
Refactor running of unit tests

### LINKED ISSUE #308: UnitTest_Engine: CheckTest method returns differences when methods return IEnumberables
#### Description:

As in header.
What's more, the regenerated test returns `Equal` on creation (it can't do anything else if it's regenerated), but turns into `Difference` after serialisation. (Needs serialising the file to check deserialisation)
image

#### Steps to reproduce:

See the test file.

#### Expected behaviour:

Should return `Equal`

#### Test file(s):

[Here](https://burohappold.sharepoint.com/:f:/s/BHoM/EpeZZx477uhBlPJFgxiavh4BR7AzX95e2ucOFDVx1H3q0A?e=81pf85) under the name FillCollinearGaps.gh.

### LINKED ISSUE #318: Allow for testing a method that returns nothing
#### Description:
This would be useful for situations like in the Geometry_Engine intersection methods where I think we need and should be able to test for cases when the method is expected to return nothing. 

For example no intersection between two lines could return nothing:
Screenshot 2021-01-21 152709
Plug those inputs to the unit test and check it:
Screenshot 2021-01-21 152900

### LINKED ISSUE #319: UnitTest_Engine: Refactor result of running unit test to be using the new TestResult class
#### Broken rules:

Result of running unit tests are not conforming to the newly added TestResult class

#### Suggestions to restore compliance:

Fix above

### LINKED ISSUE #320: UnitTest_Engine: Add CheckTest method that accepts a Dataset of UnitTests and runs through all the tests
#### Description:

Addition of a method that accepts a Dataset containing UnitTests and runs through all of the tests in it.

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### Issues addressed by this PR

Closes #319 
Closes #320 
Closes #318 
Closes #308 

 

Refactor of the UnitTest checking methods to return the new TestResult object.

 ### Test files

https://burohappold.sharepoint.com/:f:/s/BHoM/EkatzhFSfEFFkFGfP13IgTIBXoY5p72_0yXfdqo4kXbYBg?e=vKkrac

 ### Changelog

 ### Additional comments

@adecler I also had to update some of your methods in the Test_Engine as well as in the TestRunner to get the code to compile. That might need some additional tweaks.

### CODE REVIEW

File: TestRunner/TestRunner.csproj [hunk 1]
```diff
@@ -33,6 +33,11 @@
     <WarningLevel>4</WarningLevel>
   </PropertyGroup>
   <ItemGroup>
+    <Reference Include="BHoM">
+      <SpecificVersion>False</SpecificVersion>
+      <HintPath>C:\ProgramData\BHoM\Assemblies\BHoM.dll</HintPath>
+      <Private>False</Private>
+    </Reference>
```
Comment (FraserGreenroyd) [inline TestRunner/TestRunner.csproj:40]: Why was this needed? We didn't change anything except in Test oM in the end didn't we?

File: Test_Engine/Query/FullMessage.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
+﻿/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2019, the respective contributors. All rights reserved.
```
Comment (FraserGreenroyd) [inline Test_Engine/Query/FullMessage.cs:3]:
```suggestion
 * Copyright (c) 2015 - 2021, the respective contributors. All rights reserved.
```

File: Test_Engine/Query/FullMessage.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Returns a full concatenated message for the test results and its inner result to the target max depth, only giving messages worse or equal to the provided severity level.")]
+        [Input("result", "The TestResult to give the full message for.")]
+        [Input("maxDepth", "Maximum level of recursiveness for inner TestInformation.")]
+        [Input("minSeverity", "The minimum level of severity of the information to report back.")]
+        [Output("message", "Full message for the TestResult")]
+        public static string FullMessage(this TestResult result, int maxDepth = 3, TestStatus minSeverity = TestStatus.Warning)
```
Comment (FraserGreenroyd) [inline Test_Engine/Query/FullMessage.cs:45]: Following our discussion, could we default to a `Pass` on the severity, so the default is to return everything? Or is there a reason for default to `Warning`? Can be happy to keep as is if there's a reason for it :smile:

File: Test_Engine/Query/IsEqualOrMoreSevere.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
+﻿/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2019, the respective contributors. All rights reserved.
```
Comment (FraserGreenroyd) [inline Test_Engine/Query/IsEqualOrMoreSevere.cs:3]:
```suggestion
 * Copyright (c) 2015 - 2021, the respective contributors. All rights reserved.
```

File: Test_Engine/Query/IsEqualOrMoreSevere.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace BH.Engine.Test
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Checks if a TestStatus is equally or more sever to a comparison status.")]
```
Comment (FraserGreenroyd) [inline Test_Engine/Query/IsEqualOrMoreSevere.cs:39]:
```suggestion
        [Description("Checks if a TestStatus is equally or more severe to a comparison status.")]
```

File: UnitTest_Engine/Query/FullMessage.cs [hunk 6]
```diff
@@ -0,0 +1,64 @@
+﻿/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2019, the respective contributors. All rights reserved.
```
Comment (FraserGreenroyd) [inline UnitTest_Engine/Query/FullMessage.cs:3]:
```suggestion
 * Copyright (c) 2015 - 2021, the respective contributors. All rights reserved.
```

File: UnitTest_Engine/Query/FullMessage.cs [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("")]
+        [Input("", "")]
+        [Output("", "")]
```
Comment (FraserGreenroyd) [inline UnitTest_Engine/Query/FullMessage.cs:43]: 🤔 👀 🙄

File: UnitTest_oM/Properties/AssemblyInfo.cs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("UnitTest_oM")]
+[assembly: AssemblyDescription("")]
+[assembly: AssemblyConfiguration("")]
+[assembly: AssemblyCompany("")]
+[assembly: AssemblyProduct("UnitTest_oM")]
+[assembly: AssemblyCopyright("Copyright ©  2021")]
```
Comment (FraserGreenroyd) [inline UnitTest_oM/Properties/AssemblyInfo.cs:13]: Please update to the standard version

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
Still need to do full review but initial glance with comments attached.

File: TestRunner/TestRunner.csproj (hunk 1 above)
Comment (IsakNaslundBh) [reply TestRunner/TestRunner.csproj:40]: It did not compile without it for me. Complained about a missing reference to `IObject`. And did not see any real harm in having BHoM referenced here, but might have missed something?

File: Test_Engine/Query/FullMessage.cs (hunk 3 above)
Comment (IsakNaslundBh) [reply Test_Engine/Query/FullMessage.cs:45]: Happy to default to `Pass`. My reasoning for defaulting to warning was to default to only show anything problematic, but happy to change

File: Test_Engine/Query/FullMessage.cs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+            //Return an empty string if the status is not at least as severe as the provided minimum.
+            if (!result.Status.IsEqualOrMoreSevere(minSeverity))
+                return "";
+
+            string message = result.Description + Environment.NewLine;
+            message += result.Message + Environment.NewLine;
+
+            if (maxDepth > 0 && result.Information.Count > 0)
+            {
+                message += "Inner results:" + Environment.NewLine;
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Test_Engine/Query/FullMessage.cs:56]: Thanks! Indenting is actually already happening. The reason it can not be seen in GH comes down to the panels in GH I think. IF you try copying over the text from the panel to for example notepad, it should be indented!

Changing the text the "Results at depth x" is tricky with the current implementation of the recursive setup. Could ofc be refactored to allow for it, but kind of like the current flow..

Would you mind checking if the indentation in something other than a GH Panel looks ok for you?

File: UnitTest_Engine/Compute/CheckTest.cs [hunk 10]
```diff
@@ -46,72 +49,190 @@ public static partial class Compute
...
+            return testResult;
+        }
+
+        /***************************************************/
+        /**** Private Methods                           ****/
+        /***************************************************/
+
+        private static TestResult CheckTest(MethodBase method, UT.TestData data, int index)
+        {
+            string description = "Data: " + (!string.IsNullOrWhiteSpace(data.Name) ? $"name: {data.Name}," : "") + $"index: {index}";
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply UnitTest_Engine/Compute/CheckTest.cs:132]: Makes sense. WIll update!

File: Test_Engine/Query/FullMessage.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (rolyhudson) [reply Test_Engine/Query/FullMessage.cs:56]: OK indenting looks good in sublime - an we get a line space!
Screenshot 2021-01-26 140128
I can see the difficulty in passing around the current depth - would be nice but not essential.

[the remark this answers is not in the document]
Comment (rolyhudson) [reply Test_Engine/Query/FullMessage.cs:56]: Also I note the indenting also indicates the depth of the Inner results so I am happy to discard the proposal of enumerating depth.
Screenshot 2021-01-26 140735

File: Test_Engine/Query/FullMessage.cs (hunk 3 above)
Comment (FraserGreenroyd) [reply Test_Engine/Query/FullMessage.cs:45]: I feel like showing everything by default works well, with different applications choosing their level as appropriate?

**Comment by rolyhudson** [comment]
Noting a couple of edge cases of unexpected behaviour. First multi outputs
Screenshot 2021-01-26 142217 
[test script](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/Test_Toolkit/%23321-RefactorRunUnitTest/%23321-MultiOutputMethod.gh?csf=1&web=1&e=w1G2S4)
Then odd one with SplitAtPoints outputs from the testdata not the same as the method itself:
Screenshot 2021-01-26 151722 [test script](https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/Test_Toolkit/%23321-RefactorRunUnitTest/%23321-SplitAtPoints.gh?csf=1&web=1&e=vmwREd)

**Comment by IsakNaslundBh** [comment]
> Noting a couple of edge cases of unexpected behaviour. First multi outputs

This I will have to look into. This Seems like something going wrong somewhere in the running.

> Then odd one with SplitAtPoints outputs from the testdata not the same as the method itself:
This I am not sure actually is a problem. The Output data is a list in itself, where each item correspond to the position in the list of outputs. For most methods this will be a single item, but for multi outputs this should be a list.

What has happened with the split at points method is that the first item in the output list is a list, and that the explode is failing to recognise this. Hence, the first item is returned as what it is, as a list object, rather than what we are used to see in GH.

As long as this method is running fine and being checked fine, this is not an issue with the Test_Toolkit or the UnitTest. If something needs to be changed, it would need to be in the explode component in BHoM_UI.

**Comment by rolyhudson** [comment]
> This I am not sure actually is a problem. The Output data is a list in itself, where each item correspond to the position in the list of outputs. For most methods this will be a single item, but for multi outputs this should be a list.

Yes sorry a mistake by me - not an issue with the Test_Toolkit.

**Comment by IsakNaslundBh** [comment]
@rolyhudson found the bug for running multiOutput methods. Have fixed it now, should hopefully be resolved.

File: UnitTest_Engine/Query/FullMessage.cs (hunk 7 above)
Comment (IsakNaslundBh) [reply UnitTest_Engine/Query/FullMessage.cs]: 👀

**Comment by rolyhudson** [comment]
> @rolyhudson found the bug for running multiOutput methods. Have fixed it now, should hopefully be resolved.

Confirmed working as expected now. Nice work.

**Review by LMarkowski — COMMENTED** [review]
I confirm that everything works as expected. Really like those changes!

Found two small slips in the code, but nothing requiring necessary change. Therefore I leave review as a comment.

File: Test_Engine/Create/FailResult.cs [hunk 11]
```diff
@@ -41,9 +42,9 @@ public static partial class Create
 
         public static TestResult FailResult(string errorMessage, string description = "", bool isCritical = false)
         {
-            ResultStatus status = isCritical ? ResultStatus.CriticalFail : ResultStatus.Fail;
-            List<Event> events = new List<Event> { new Event { Type = EventType.Error, Message = errorMessage } };
-            return new TestResult(status, events, description);
+            TestStatus status = isCritical ? TestStatus.Error : TestStatus.Warning;
+            //List<Event> events = new List<Event> { new Event { Type = EventType.Error, Message = errorMessage } };
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Test_Engine/Create/FailResult.cs:46]: Think it might be. This is part of code that @adecler will update in an upcoming PR. I just tweaked it to compile, and while doing so, tried to keep things I could not properly update commented out. Agree that we should never leave commented out code in the general case, but I think this should be ok as an exception.

File: UnitTest_Engine/Query/FullMessage.cs [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Returns a full concatenated message for the ComparisonDifference, only giving messages worse or equal to the provided severity level.")]
+        [Input("result", "The ComparisonDifference to give the full message for.")]
+        [Input("maxDepth", "Maximum level of recursiveness for inner TestInformation. Not in use for this object type.")]
+        [Input("minSeverity", "The minimum level of severity of the information to report back. Returns an empty string if the ComparisonDifference does not pass this check.")]
+        [Output("message", "Full message for the ComparisonDifference")]
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply UnitTest_Engine/Query/FullMessage.cs:45]: 👀

File: Test_Engine/Query/FullMessage.cs [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+            return message;
+        }
+
+        /***************************************************/
+        /**** Public Methods - Interface                ****/
+        /***************************************************/
+
+        public static string IFullMessage(this ITestInformation result, int maxDepth = 3, TestStatus minSeverity = TestStatus.Pass)
+        {
+            return FullMessage(result, maxDepth, minSeverity);
```
Comment (adecler) [inline Test_Engine/Query/FullMessage.cs:82]: There is no `as dynamic` here so what is the point of this method ?

File: TestRunner/Program.cs [hunk 14]
```diff
@@ -62,7 +62,7 @@ static void Main(string[] args)
                         TestResult result = method.Invoke(null, new object[] { }) as TestResult;
 
                         Console.WriteLine(result.Description + " --> " + result.Status.ToString());
-                        foreach (Event e in result.Events.OrderBy(x => x.Message))
+                        foreach (Event e in result.Information.OrderBy(x => x.Message))
```
Comment (adecler) [inline TestRunner/Program.cs:65]: `e` is now an `ITestInformation`, not an `Event`

File: Test_Engine/Create/FailResult.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (adecler) [reply Test_Engine/Create/FailResult.cs:46]: It can be safely deleted 👍

**Review by adecler — CHANGES_REQUESTED** [review]
Overall looking good. Just a few little things that need correcting.

File: Test_Engine/Query/FullMessage.cs (hunk 13 above)
Comment (IsakNaslundBh) [reply Test_Engine/Query/FullMessage.cs:82]: 🤦 you are right. Did not spot it as it was giving the same results due to the `RunExtensionMethod` in the fallback method. Will fix.

**Review by adecler — COMMENTED** [review]
I am now happy. Will leave others to give the final verdict as I see other requests for change.

**Review by adecler — DISMISSED** [review]

**Comment by IsakLarbornBH** [comment]
Slightly off topic but:

> @adecler I also had to update some of your methods in the Test_Engine as well as in the TestRunner to get the code to compile. That might need some additional tweaks.

Not just you @IsakNaslundBh - I tried for a while to build Test_Toolkit from master but it was just failing every time no matter what I did. Found this PR while looking for a solution and this branch builds perfectly, no errors. Not sure what that's all about.

File: UnitTest_Engine/Compute/Run.cs [hunk 15]
```diff
@@ -104,7 +104,7 @@ public static partial class Compute
             {
                 string message = "Failed to run with the given inputs";
                 if (e is NotImplementedException || e.InnerException is NotImplementedException)
-                    message = "The method is not implements";
+                    message = "The method is not implemented.";
                 else if (e is RuntimeBinderException || e.InnerException is RuntimeBinderException)
                     message = "The method with the correct input types cannot be found";
```
Comment (FraserGreenroyd) [inline UnitTest_Engine/Compute/Run.cs:109]:
```suggestion
                    message = "The method with the correct input types cannot be found.";
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
Very very minor

**Review by FraserGreenroyd — APPROVED** [review]
This LGTM tack sa mycket @IsakNaslundBh :smile: thanks for the call just now too, really good to bottom out some of our thoughts on the framework around this as we move forward :smile:

**Comment by adecler** [comment]
> Not sure what that's all about.

[This PR](https://github.com/BHoM/BHoM/pull/1154) temporarily broke some parts of the Test toolkit and the Verification solutions in CI/CD. That's what it's all about.
